src/System.Private.CoreLib/src/System/Threading/Timer.Unix.cs

```diff
@@ -13,109 +14,128 @@ namespace System.Threading
+        {
+            Debug.Assert(s_lock.IsAcquired);
+            Debug.Assert(s_appDomainTimers == null);
+
+            var appDomainTimers = new List<AppDomainTimer>(Instances.Length);
+
+            RuntimeThread timerThread = RuntimeThread.Create(TimerThread);
+            timerThread.IsBackground = true;
+            timerThread.Start();
+
+            // Do this after creating the thread in case thread creation fails so that it will try again next time
+            s_appDomainTimers = appDomainTimers;
+            return appDomainTimers;
         }
 
         private bool SetTimer(uint actualDuration)
         {
-            // Note: AutoResetEvent.WaitOne takes an Int32 value as a timeout.
-            // The TimerQueue code ensures that timer duration is not greater than max Int32 value
-            Debug.Assert(actualDuration <= (uint)int.MaxValue);
-            s_nextTimerDuration = (int)actualDuration;
-
-            // If this is the first time the timer is set then we need to create a thread that
-            // will manage and respond to timer requests. Otherwise, simply signal the timer thread
-            // to notify it that the timer duration has changed.
-            if (s_timerEvent == null)
-            {
-                s_timerEvent = new AutoResetEvent(false);
-                RuntimeThread thread = RuntimeThread.Create(TimerThread, 0);
-                thread.IsBackground = true; // Keep this thread from blocking process shutdown
-                thread.Start();
-            }
-            else
+            Debug.Assert((int)actualDuration >= 0);
+            var timer = new AppDomainTimer(_id, TickCount + (int)actualDuration);
+            using (LockHolder.Hold(s_lock))
             {
-                s_timerEvent.Set();
+                List<AppDomainTimer> timers = s_appDomainTimers;
+                if (timers == null)
+                {
+                    timers = InitializeAppDomainTimerManager_Locked();
+                }
+
+                timers.Add(timer);
             }
 
+            s_timerEvent.Set();
             return true;
         }
 
-
         /// <summary>
         /// This method is executed on a dedicated a timer thread. Its purpose is
-        /// to handle timer request and notify the TimerQueue when a timer expires.
+        /// to handle timer requests and notify the TimerQueue when a timer expires.
         /// </summary>
-        private void TimerThread()
+        private static void TimerThread()
         {
-            int currentTimerInterval;
-
-            // Get wait time for the next timer
-            currentTimerInterval = Interlocked.Exchange(ref s_nextTimerDuration, Timeout.Infinite);
+            AutoResetEvent timerEvent = s_timerEvent;
+            Lock lok = s_lock;
+            List<AppDomainTimer> timers;
+            using (LockHolder.Hold(lok))
+            {
+                timers = s_appDomainTimers;
+            }
```

Comment:
Why have the lock here? If the initialization in order in `InitializeAppDomainTimerManager_Locked` was changed then it could be simple assignment.